### PR TITLE
Events in IR generator.

### PR DESCRIPTION
**Pull request by chriseth**
Part of #8343.

### CODE REVIEW

File: libdevcore/StringUtils.h [hunk 1]
```diff
@@ -74,6 +74,22 @@ std::string joinHumanReadable
...
+/// unless the collection is empty.
+template<class T>
+std::string joinHumanReadablePrefixed
+(
+	T const& _list,
+	std::string const& _separator = ", ",
+	std::string const& _lastSeparator = ""
+)
+{
+	if (begin(_list) == end(_list))
```
Comment (ekpyron) [inline libdevcore/StringUtils.h:87]: I wonder whether this should be ``std::begin`` and ``std::end``...
It probably only works without ``std::`` due to argument-dependent-lookup or something similar in the current use cases, but it won't always.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 2]
```diff
@@ -963,6 +963,37 @@ string YulUtilFunctions::validatorFunction(Type const& _type, bool _revertOnFail
...
+	return m_functionCollector->createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(<variables>) -> hash {
+				let pos := mload(<freeMemoryPointer>)
+				let end := <packedEncode>(pos <comma> <variables>)
+				hash := keccak256(pos, sub(end, pos))
+			}
+		)");
+		templ("functionName", functionName);
+		templ("variables", suffixedVariableNameList("var_", 1, 1 + sizeOnStack));
```
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp:989]: We could add something like ``joinHumanReadablePrefixed`` for ``suffixedVariableNameList`` as well to get rid of all the ``<comma>`` stuff - but not necessarily in this PR.

**Review by ekpyron — COMMENTED** [review]
None of the tests actually involve reference types or did I overlook those? If there are none, we should add some...

File: libdevcore/StringUtils.h (hunk 1 above)
Comment (chriseth) [reply libdevcore/StringUtils.h:87]: Yeah, I copied that from the function above...

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:989]: But we shouldn't call it `suffixedVariableNameListPrefixed` ;)

Another option would be to add a function `commaIfNeeded(x) { return x.empty() ? "" : ", " + x; }`

**Comment by chriseth** [comment]
Reference types are not yet implemented, that's the reason.

**Comment by chriseth** [comment]
There are tests that use reference types with events, but I did not include them for the yul pipeline.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**